libpod/networking_linux.go

```diff
@@ -301,206 +302,218 @@ func (r *RootlessCNI) Cleanup(runtime *Runtime) error {
+				resolveIP, err = GetSlirp4netnsDNS(cidr)
 				if err != nil {
-					return nil, errors.Wrap(err, "could not create rootless-cni netns directory")
+					return nil, errors.Wrapf(err, "failed to determine slirp4netns DNS address from cidr: %s", cidr.String())
 				}
-			} else {
-				// return a error if we could not get the namespace and should no create one
-				return nil, errors.Wrap(err, "error getting rootless cni network namespace")
 			}
-		}
+			conf, err := resolvconf.Get()
+			if err != nil {
+				return nil, err
+			}
+			conf, err = resolvconf.FilterResolvDNS(conf.Content, netOptions.enableIPv6, true)
+			if err != nil {
+				return nil, err
+			}
+			searchDomains := resolvconf.GetSearchDomains(conf.Content)
+			dnsOptions := resolvconf.GetOptions(conf.Content)
+			nameServers := resolvconf.GetNameservers(conf.Content)
 
-		// The CNI plugins need access to iptables in $PATH. As it turns out debian doesn't put
-		// /usr/sbin in $PATH for rootless users. This will break rootless cni completely.
-		// We might break existing users and we cannot expect everyone to change their $PATH so
-		// lets add /usr/sbin to $PATH ourselves.
-		path = os.Getenv("PATH")
-		if !strings.Contains(path, "/usr/sbin") {
-			path = path + ":/usr/sbin"
-			os.Setenv("PATH", path)
-		}
+			_, err = resolvconf.Build(filepath.Join(cniDir, "resolv.conf"), append([]string{resolveIP.String()}, nameServers...), searchDomains, dnsOptions)
+			if err != nil {
+				return nil, errors.Wrap(err, "failed to create rootless cni resolv.conf")
+			}
 
-		rootlessCNINS = &RootlessCNI{
-			ns:   ns,
-			dir:  cniDir,
-			lock: lock,
+			// create cni directories to store files
+			// they will be bind mounted to the correct location in a extra mount ns
+			err = os.MkdirAll(filepath.Join(cniDir, strings.TrimPrefix(persistentCNIDir, "/")), 0700)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not create rootless-cni var directory")
+			}
+			runDir := filepath.Join(cniDir, "run")
+			err = os.MkdirAll(runDir, 0700)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not create rootless-cni run directory")
+			}
+			// relabel the new run directory to the iptables /run label
+			// this is important, otherwise the iptables command will fail
+			err = label.Relabel(runDir, "system_u:object_r:iptables_var_run_t:s0", false)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not create relabel rootless-cni run directory")
+			}
+			// create systemd run directory
+			err = os.MkdirAll(filepath.Join(runDir, "systemd"), 0700)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not create rootless-cni systemd directory")
+			}
+			// create the directory for the netns files at the same location
+			// relative to the rootless-cni location
+			err = os.MkdirAll(filepath.Join(cniDir, nsDir), 0700)
+			if err != nil {
+				return nil, errors.Wrap(err, "could not create rootless-cni netns directory")
+			}
+		} else {
```

Comment:
Pre-existing to this PR: I find the code a bit too nested.  I think we can reduce the scopes/branches a bit more.

```golang
if err != nil {
   if !new {
   return nil, errors.Wrap(...)
  }
  // handle new == true
}
```

Reply:
done